### ISSUE TITLE
Support py_volcano platform

### ISSUE THREAD

**Issue by zhengchenyu** [intelligent-machine-learning/dlrover#1559]
Some companies have used volcano as the default batch scheduler. Elasticjob uses the default scheduler. If there are multiple schedulers working at the same time, it may cause inconsistency. Volcano has useful features such as gang scheduling and queue management. Therefore, it is necessary to support the dlrover job runing on Volcano.
We can create vcjob(but not elasticjob), tasks are master and worker. Use volcano to scaling.

> Note: Go version [volcano batch scheduler](https://github.com/intelligent-machine-learning/dlrover/blob/master/go/master/pkg/batchscheduler/volcano.go) have already exists. So I name this py_volcano platform.

How to do?
* Add dlrover plugin for volcano 
* Create vcjob for volcano

**Comment by BalaBalaYi** [comment]
gang scheduling is wip but not by volcano(by kueue) 

Are you looking to participate and work on the integration with Volcano? From a development perspective, it’s likely that the implementation will need to adapt to Volcano’s CRD (Custom Resource Definition).  BTW, you can wait for some updated information regarding our collaboration with Kueue next month.

**Comment by zhengchenyu** [comment]
@BalaBalaYi 
Volcano and kueue are both options for gang scheduling.
About integration with Volcano. I think I have two proposal:
* (1) We still use elasticjob. But In elasticjob controller, we watch elasticjob, then create podgroup for volcano.
* (2) We do not use elasticjob, but use vcjob of volcano.

I temporarily used the simpler proposal (2) to verify and submitted a [draft pr](https://github.com/intelligent-machine-learning/dlrover/pull/1560).

**Comment by zhengchenyu** [comment]
* Add dlrover plugin for volcano
example plugin for dlrover are:
```
package dlrover

import (
	"context"
	"flag"
	"fmt"
	corev1 "k8s.io/api/core/v1"
	v1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/klog/v2"
	"math"
	"strconv"
	batch "volcano.sh/apis/pkg/apis/batch/v1alpha1"
	"volcano.sh/volcano/pkg/controllers/job/helpers"
	pluginsinterface "volcano.sh/volcano/pkg/controllers/job/plugins/interface"
	"volcano.sh/volcano/pkg/scheduler/api"
)

const (
	// DlroverPluginName is the name of the plugin
	DlroverPluginName = "dlrover"
	// DefaultPort is the default port for dlrover master
	DefaultPort = 50001
	// DefaultMaster is the default task name of master host
	DefaultMaster = "master"
	// DefaultWorker is the default task name of worker host
	DefaultWorker = "worker"

	// LabelAppNameKey is the key of application name in the cluster.
	LabelAppNameKey = "app.kubernetes.io/name"
	// ElasticJobAppName is the application name of dlrover elasticjob.
	ElasticJobAppName = "elasticjob"
	// LabelJobNameKey is the label key of the elasticjob name.
	LabelJobNameKey = "elasticjob.dlrover/name"
	// LabelReplicaTypeKey is the key of ReplicaType in labels.
	LabelReplicaTypeKey  = "elasticjob.dlrover/replica-type"
	JobMasterReplicaType = "dlrover-master"
	WORKER               = "worker"
	// LabelReplicaIndexKey is the key of ReplicaIndex in labels.
	// For volcano, replica index = rank index = pod index
	LabelReplicaIndexKey = "elasticjob.dlrover/replica-index"
	// LabelRankIndexKey is the key of rankIndex.
	LabelRankIndexKey = "elasticjob.dlrover/rank-index"
	// JobMasterReplicaType is the type for DLRover ElasticJob Master replica.
	LabelRelaunchCount = "elasticjob.dlrover/relaunch-count"

	// Environment variables
	JOB_NAME                    = "ELASTIC_JOB_NAME"
	JOB_UID                     = "JOB_UID"
	GRPC_ENABLE_FORK            = "GRPC_ENABLE_FORK_SUPPORT"
	NODE_TYPE                   = "NODE_TYPE"
	NODE_ID                     = "NODE_ID"
	NODE_NUM                    = "NODE_NUM"
	NODE_RANK                   = "NODE_RANK"
	RANK                        = "RANK"
	WORLD_SIZE                  = "WORLD_SIZE"
	MAX_NODE                    = "MAX_NODE"
	MIN_NODE                    = "MIN_NODE"
	DLROVER_MASTER_ADDR         = "DLROVER_MASTER_ADDR"
	DLROVER_MASTER_SERVICE_TYPE = "DLROVER_MASTER_SERVICE_TYPE"
	POD_NAME                    = "POD_NAME"

	CONFIGMAP_SUFFIX = "-dlrover-conf"
)

type dlroverPlugin struct {
	arguments            []string
	clientset            pluginsinterface.PluginClientset
	masterName           string
	workerName           string
	serviceType          string
	priority             int
	port                 int
	optimizeMode         string
	distributionStrategy string
}

// New creates mpi plugin.
func New(client pluginsinterface.PluginClientset, arguments []string) pluginsinterface.PluginInterface {
	dp := dlroverPlugin{arguments: arguments, clientset: client}
	dp.addFlags()
	return &dp
}

func (dp *dlroverPlugin) addFlags() {
	flagSet := flag.NewFlagSet(dp.Name(), flag.ContinueOnError)
	flagSet.IntVar(&dp.port, "port", DefaultPort, "service port")
	flagSet.StringVar(&dp.masterName, "master", DefaultMaster, "name of master role task")
	flagSet.StringVar(&dp.workerName, "worker", DefaultWorker, "name of worker role task")
	flagSet.StringVar(&dp.serviceType, "serviceType", "grpc", "name of service type")
	flagSet.IntVar(&dp.priority, "priority", 100, "priority of worker")
	flagSet.StringVar(&dp.optimizeMode, "optimizeMode", "single-job", "optimize mode of elastic job")
	flagSet.StringVar(&dp.distributionStrategy, "distributionStrategy", "AllreduceStrategy", "distribution strategy of elastic job")

	if err := flagSet.Parse(dp.arguments); err != nil {
		klog.Errorf("plugin %s flagset parse failed, err: %v", dp.Name(), err)
	}
}

func (dp *dlroverPlugin) Name() string {
	return DlroverPluginName
}

func (dp *dlroverPlugin) OnPodCreate(pod *v1.Pod, job *batch.Job) error {
	taskType := helpers.GetTaskKey(pod)
	if taskType == dp.masterName {
		return dp.OnMaterPodCreate(pod, job)
	} else if taskType == dp.workerName {
		return dp.OnWorkerPodCreate(pod, job)
	} else {
		fmt.Errorf("invalid task type %v for job %v", taskType, job.Name)
	}
	return nil
}

func (dp *dlroverPlugin) OnMaterPodCreate(pod *v1.Pod, job *batch.Job) error {
	// Set labels
	pod.Labels[LabelAppNameKey] = ElasticJobAppName
	pod.Labels[LabelJobNameKey] = job.Name
	pod.Labels[LabelReplicaTypeKey] = JobMasterReplicaType
	pod.Labels[LabelReplicaIndexKey] = "0"
	// The priority of the master is always 1 greater than that of the worker.
	pod.Annotations[api.TaskPriorityAnnotation] = strconv.Itoa(dp.priority + 1)
	// Set environment variables
	envVars := []v1.EnvVar{}
	envVars = append(envVars, v1.EnvVar{Name: JOB_NAME, Value: job.Name})
	for i, _ := range pod.Spec.Containers {
		pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, envVars...)
	}
	return nil
}

func (dp *dlroverPlugin) OnWorkerPodCreate(pod *v1.Pod, job *batch.Job) error {
	// Calculate master address and the replica of worker.
	masterIndex := helpers.GetTaskIndexUnderJob(dp.masterName, job)
	if masterIndex == -1 {
		klog.Errorf("job %v doesn't have task %v", job.Name, dp.masterName)
		return nil
	}
	workerIndex := helpers.GetTaskIndexUnderJob(dp.workerName, job)
	if workerIndex == -1 {
		klog.Errorf("job %v doesn't have task %v", job.Name, dp.workerName)
		return nil
	}
	replicasStr := strconv.Itoa(int(job.Spec.Tasks[workerIndex].Replicas))
	masterAddr := dp.generateMasterAddr(job.Spec.Tasks[masterIndex], job.Name)

	// Set labels
	pod.Labels["app"] = "dlrover"
	pod.Labels[LabelJobNameKey] = job.Name
	pod.Labels[LabelReplicaTypeKey] = WORKER
	pod.Labels[LabelReplicaIndexKey] = pod.Labels[batch.TaskIndex]
	pod.Labels[LabelRankIndexKey] = pod.Labels[batch.TaskIndex]
	// disable relaunch
	pod.Labels[LabelRelaunchCount] = strconv.Itoa(math.MaxInt)
	pod.Annotations[api.TaskPriorityAnnotation] = strconv.Itoa(dp.priority)

	// Set environment variables
	envVars := []v1.EnvVar{}
	envVars = append(envVars, v1.EnvVar{
		Name: "POD_IP",
		ValueFrom: &v1.EnvVarSource{
			FieldRef: &v1.ObjectFieldSelector{
				FieldPath: "status.podIP",
			},
		},
	}, v1.EnvVar{
		Name: "NODE_IP",
		ValueFrom: &v1.EnvVarSource{
			FieldRef: &v1.ObjectFieldSelector{
				FieldPath: "status.hostIP",
			},
		},
	}, v1.EnvVar{
		Name:  JOB_NAME,
		Value: job.Name,
	}, v1.EnvVar{
		Name:  JOB_UID,
		Value: string(job.UID),
	}, v1.EnvVar{
		Name:  GRPC_ENABLE_FORK,
		Value: "false",
	}, v1.EnvVar{
		Name:  NODE_TYPE,
		Value: WORKER,
	}, v1.EnvVar{
		Name:  NODE_ID,
		Value: pod.Labels[batch.TaskIndex],
	}, v1.EnvVar{
		Name:  NODE_NUM,
		Value: replicasStr,
	}, v1.EnvVar{
		Name:  NODE_RANK,
		Value: pod.Labels[batch.TaskIndex],
	}, v1.EnvVar{
		Name:  WORLD_SIZE,
		Value: replicasStr,
	}, v1.EnvVar{
		Name:  RANK,
		Value: pod.Labels[batch.TaskIndex],
	}, v1.EnvVar{
		Name:  MAX_NODE,
		Value: replicasStr,
	}, v1.EnvVar{
		Name: MIN_NODE,
		// Subtract the master node
		Value: strconv.Itoa(int(max(job.Spec.MinAvailable-
			job.Spec.Tasks[masterIndex].Replicas, 1))),
	}, v1.EnvVar{
		Name:  DLROVER_MASTER_ADDR,
		Value: fmt.Sprintf("%s:%v", masterAddr, dp.port),
	}, v1.EnvVar{
		Name:  DLROVER_MASTER_SERVICE_TYPE,
		Value: dp.serviceType,
	}, v1.EnvVar{
		Name: POD_NAME,
		ValueFrom: &v1.EnvVarSource{
			FieldRef: &v1.ObjectFieldSelector{
				FieldPath: "metadata.name",
			},
		},
	})
	for i, _ := range pod.Spec.Containers {
		pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, envVars...)
	}
	return nil
}

func (dp *dlroverPlugin) OnJobAdd(job *batch.Job) error {
	if job.Status.ControlledResources["plugin-"+dp.Name()] == dp.Name() {
		return nil
	}
	job.Status.ControlledResources["plugin-"+dp.Name()] = dp.Name()
	if err := dp.createConfigMapIfNotExists(job); err != nil {
		return err
	}
	dp.updatePriorityIfnecessary(job)
	return nil
}

func (dp *dlroverPlugin) OnJobDelete(job *batch.Job) error {
	if job.Status.ControlledResources["plugin-"+dp.Name()] != dp.Name() {
		return nil
	}
	delete(job.Status.ControlledResources, "plugin-"+dp.Name())
	return nil
}

func (dp *dlroverPlugin) OnJobUpdate(job *batch.Job) error {
	return nil
}

func (dp *dlroverPlugin) generateMasterAddr(task batch.TaskSpec, jobName string) string {
	hostName := task.Template.Spec.Hostname
	subdomain := task.Template.Spec.Subdomain
	if len(hostName) == 0 {
		hostName = helpers.MakePodName(jobName, task.Name, 0)
	}
	if len(subdomain) == 0 {
		subdomain = jobName
	}

	host := hostName + "." + subdomain
	return host
}

func (dp *dlroverPlugin) createConfigMapIfNotExists(job *batch.Job) error {
	name := job.Name + CONFIGMAP_SUFFIX
	if _, err := dp.clientset.KubeClients.CoreV1().ConfigMaps(job.Namespace).Get(
		context.TODO(), name, metav1.GetOptions{}); err == nil {
		klog.Infof("ConfigMap %v/%v already exists, skipping creation.\n", job.Namespace, name)
		return nil
	}

	controller := true
	configMap := &corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{
			Name:      name,
			Namespace: job.Namespace,
			OwnerReferences: []metav1.OwnerReference{
				{
					APIVersion: "batch.volcano.sh/v1alpha1",
					Kind:       "Job",
					Name:       job.Name,
					UID:        job.UID,
					Controller: &controller,
				},
			},
		},
		Data: map[string]string{
			"optimizeMode":         dp.optimizeMode,
			"distributionStrategy": dp.distributionStrategy,
		},
	}

	if _, err := dp.clientset.KubeClients.CoreV1().ConfigMaps(job.Namespace).Create(
		context.TODO(), configMap, metav1.CreateOptions{}); err != nil {
		return err
	}
	klog.Infof("Created ConfigMap: %v/%v\n", job.Namespace, name)
	return nil
}

func (dp *dlroverPlugin) updatePriorityIfnecessary(job *batch.Job) {
	if job.Spec.PriorityClassName != "" {
		if pc, err := dp.clientset.KubeClients.SchedulingV1().PriorityClasses().Get(
			context.TODO(), job.Spec.PriorityClassName, metav1.GetOptions{}); err == nil {
			dp.priority = int(pc.Value)
		}
	}
}

```
* Create vcjob for volcano

example vcjob crd:

```
apiVersion: batch.volcano.sh/v1alpha1
kind: Job
metadata:
  name: dlrover-volcano-job
  namespace: dlrover
spec:
  minAvailable: 5
  schedulerName: volcano
  priorityClassName: high-priority
  plugins:
    env: []
    svc: []
    dlrover: [--port=50001]
  maxRetry: 5
  queue: default
  tasks:
    - name: "master"
      replicas: 1
      policies:
      - event: PodEvicted
        action: RestartJob
      - event: TaskCompleted
        action: CompleteJob
      - event: TaskFailed
        action: RestartJob
      template:
        metadata:
          name: master
        spec:
          containers:
            - name: master
              image: ${XX_IMAGE}:${XX_TAG}
              imagePullPolicy: Always
              command:
                - /bin/bash
                - -c
                - "python -m dlrover.python.master.main --namespace dlrover --platform py_volcano \
                   --job_name ${ELASTIC_JOB_NAME} --port 50001"
              resources:
                requests:
                  cpu: "2"
                  memory: 8Gi
          restartPolicy: Always
    - name: "worker"
      replicas: 6
      template:
        metadata:
          name: worker
        spec:
          containers:
            - name: worker
              image:${XX_IMAGE}:${XX_TAG}
              imagePullPolicy: Always
              command:
                - /bin/bash
                - -c
                - " dlrover-run --nproc_per_node=2 --max_restarts=3 --rdzv_conf='join_timeout=600' \
                   /xxx/cnn_train.py --num_epochs 10000 --fixed_batch_size=false"
              resources:
                requests:
                  cpu: "2"
                  memory: 8Gi
                  # nvidia.com/gpu: 1
          restartPolicy: Always

```

**Comment by workingloong** [comment]
> [@BalaBalaYi](https://github.com/BalaBalaYi) Volcano and kueue are both options for gang scheduling. About integration with Volcano. I think I have two proposal:
> 
> * (1) We still use elasticjob. But In elasticjob controller, we watch elasticjob, then create podgroup for volcano.
> * (2) We do not use elasticjob, but use vcjob of volcano.
> 
> I temporarily used the simpler proposal (2) to verify and submitted a [draft pr](https://github.com/intelligent-machine-learning/dlrover/pull/1560).

Proposal (2) is a lightweight deployment solution compared with (1). However, proposal (1) can automatically cordon the fault nodes.

**Comment by ldengjie** [comment]
> gang scheduling is wip but not by volcano(by kueue)
> 
> Are you looking to participate and work on the integration with Volcano? From a development perspective, it’s likely that the implementation will need to adapt to Volcano’s CRD (Custom Resource Definition). BTW, you can wait for some updated information regarding our collaboration with Kueue next month.

Hi, thank you for your hard work. May I ask about the progress related to Kueue? When will it be approximately ready for release and use?
